### PR TITLE
GH-84783: Make the slice object hashable

### PR DESCRIPTION
**Pull request by furkanonder**
* Issue: gh-84783

### CODE REVIEW

File: Objects/sliceobject.c [hunk 1]
```diff
@@ -628,6 +628,42 @@ slice_traverse(PySliceObject *v, visitproc visit, void *arg)
     return 0;
 }
 
+/* code based on tuplehash() of Objects/tupleobject.c */
```
Comment (rhettinger) [inline Objects/sliceobject.c:631]: This is all nice work; however, a simpler solution will suffice.

```
def slice_hash(s):
    # Multiply by two randomly chosen primes
    return hash(s.start) * 314898481 | hash(s.stop) * 4822421 | hash(s.step)
```

**Review by rhettinger — CHANGES_REQUESTED** [review]

File: Objects/sliceobject.c (hunk 1 above)
Comment (rhettinger) [reply Objects/sliceobject.c:631]: After more thought, I think we can leave your version as-is.  If someone wants to propose a simplification and test collision rates, they can do so at a later time.  It is reasonable to stick with the known solution.

**Review by rhettinger — APPROVED** [review]

**Review by rhettinger — CHANGES_REQUESTED** [review]
We do need to add a NEWS blurb and a `versionchanged` entry to the docs.

**Comment by rhettinger** [comment]
We're getting some test failures:

```
======================================================================
[3683](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3684)
ERROR: test_sequence_del_slice (test.test_capi.test_misc.CAPITest.test_sequence_del_slice)
[3684](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3685)
----------------------------------------------------------------------
[3685](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3686)
Traceback (most recent call last):
[3686](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3687)
  File "/Users/runner/work/cpython/cpython/Lib/test/test_capi/test_misc.py", line 459, in test_sequence_del_slice
[3687](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3688)
    _testcapi.sequence_del_slice(mapping, 1, 3)
[3688](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3689)
KeyError: slice(1, 3, None)
[3689](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3690)

[3690](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3691)
======================================================================
[3691](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3692)
FAIL: test_sequence_set_slice (test.test_capi.test_misc.CAPITest.test_sequence_set_slice)
[3692](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3693)
----------------------------------------------------------------------
[3693](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3694)
Traceback (most recent call last):
[3694](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3695)
  File "/Users/runner/work/cpython/cpython/Lib/test/test_capi/test_misc.py", line 419, in test_sequence_set_slice
[3695](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3696)
    with self.assertRaises(TypeError):
[3696](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3697)
AssertionError: TypeError not raised
[3697](https://github.com/python/cpython/actions/runs/4006251040/jobs/6877559173#step:7:3698)
```

File: Objects/sliceobject.c [hunk 2]
```diff
@@ -628,6 +628,42 @@ slice_traverse(PySliceObject *v, visitproc visit, void *arg)
...
+#define _PyHASH_XXPRIME_2 ((Py_uhash_t)2246822519UL)
+#define _PyHASH_XXPRIME_5 ((Py_uhash_t)374761393UL)
+#define _PyHASH_XXROTATE(x) ((x << 13) | (x >> 19))  /* Rotate left 13 bits */
+#endif
+
+static Py_hash_t
+slicehash(PySliceObject *v)
+{
+    Py_uhash_t acc = _PyHASH_XXPRIME_5;
+#define _PyHASH_SLICE_PART(com) { \
```
Comment (rhettinger) [inline Objects/sliceobject.c:648]: It would look nicer if this macro were replaced with a `static inline function`.

File: Lib/test/test_slice.py [hunk 3]
```diff
@@ -80,10 +80,8 @@ def test_repr(self):
         self.assertEqual(repr(slice(1, 2, 3)), "slice(1, 2, 3)")
 
     def test_hash(self):
-        # Verify clearing of SF bug #800796
-        self.assertRaises(TypeError, hash, slice(5))
-        with self.assertRaises(TypeError):
-            slice(5).__hash__()
+        self.assertEqual(hash(slice(5)), slice(5).__hash__())
+        self.assertNotEqual(slice(5), slice(6))
```
Comment (OTheDev) [inline Lib/test/test_slice.py:86]: Can we add more tests here?

For example, test that `hash(slice(1, 2, []))` raises `TypeError` (to test the portion of the code where `PyObject_Hash()` fails).

Comment (OTheDev) [reply Lib/test/test_slice.py:86]: I think there should also be tests for the multiple argument forms of `slice()` (:

File: Doc/library/functions.rst [hunk 4]
```diff
@@ -1635,6 +1635,8 @@ are always available.  They are listed here in alphabetical order.
    example: ``a[start:stop:step]`` or ``a[start:stop, i]``.  See
    :func:`itertools.islice` for an alternate version that returns an iterator.
 
+   .. versionchanged:: 3.10
```
Comment (OTheDev) [inline Doc/library/functions.rst]: I don't think this can be backported?

Comment (OTheDev) [reply Doc/library/functions.rst]:
```suggestion
   .. versionchanged:: 3.12
      Slice objects are now :term:`hashable` (provided :attr:`~slice.start`,
      :attr:`~slice.stop`, and :attr:`~slice.step` are hashable).
```

File: Lib/test/test_slice.py (hunk 3 above)
Comment (furkanonder) [reply Lib/test/test_slice.py:86]: I added some test cases. Please do not hesitate to write if any other situations come to mind. Thank you so much for your wonderful feedback and contributions!

File: Lib/test/test_slice.py [hunk 5]
```diff
@@ -80,10 +80,17 @@ def test_repr(self):
         self.assertEqual(repr(slice(1, 2, 3)), "slice(1, 2, 3)")
 
     def test_hash(self):
-        # Verify clearing of SF bug #800796
-        self.assertRaises(TypeError, hash, slice(5))
+        self.assertEqual(hash(slice(5)), slice(5).__hash__())
+        self.assertEqual(hash(slice(1, 2)), slice(1, 2).__hash__())
+        self.assertEqual(hash(slice(1, 2, 3)), slice(1, 2, 3).__hash__())
+        self.assertEqual(hash((slice(4, 2), slice(2, 6))), (slice(4, 2), slice(2, 6)).__hash__())
+        self.assertNotEqual(slice(5), slice(6))
```
Comment (OTheDev) [inline Lib/test/test_slice.py]: Thanks! I suggest removing these two lines.

Comparison of slice objects and comparisons of tuple hashes can be left to `test_cmp()` and `test_tuple.py`, respectively.

Comment (furkanonder) [reply Lib/test/test_slice.py]: It make sense. Thanks!

File: Lib/test/test_slice.py [hunk 6]
```diff
@@ -80,10 +80,16 @@ def test_repr(self):
         self.assertEqual(repr(slice(1, 2, 3)), "slice(1, 2, 3)")
 
     def test_hash(self):
-        # Verify clearing of SF bug #800796
-        self.assertRaises(TypeError, hash, slice(5))
+        self.assertEqual(hash(slice(5)), slice(5).__hash__())
+        self.assertEqual(hash(slice(1, 2)), slice(1, 2).__hash__())
+        self.assertEqual(hash(slice(1, 2, 3)), slice(1, 2, 3).__hash__())
+        self.assertNotEqual(slice(5), slice(6))
```
Comment (OTheDev) [inline Lib/test/test_slice.py:86]: Why test this (line 86)? Isn't that what `test_cmp()` is for?

**Pull request merged**